### PR TITLE
fix #157: split function accounting for quotes

### LINKED ISSUE #157: bug: regexMatch() cannot parse {N,M} pattern quantifier - "invalid policy size" error
## Problem

When using the builtin function [`regexMatch()`](https://github.com/casbin/lua-casbin/blob/24bd8d63972a2b188198dfa301c7dbcf78eede31/src/util/BuiltInFunctions.lua#L173-L187), I'm not able to use the pattern quantifier `{N,M}`, because it contains a comma: it's parsed as 2 arguments and the enforcer throws an error ["invalid policy size"](https://github.com/casbin/lua-casbin/blob/24bd8d63972a2b188198dfa301c7dbcf78eede31/src/main/CoreEnforcer.lua#L423).

## Example Use Case

Here's an example for reference:

- file `model.conf`:
  
  ```conf
  [request_definition]
  r = sub, obj, act
  [policy_definition]
  p = sub, obj, act
  [policy_effect]
  e = some(where (p.eft == allow))
  [matchers]
  m = regexMatch(r.sub, p.sub) && regexMatch(r.obj, p.obj) && regexMatch(r.act, p.act)
  ```

- file `policy.csv`:
  
  ```csv
  p, test, ^/test/[a-zA-Z0-9]{1,3}$, GET
  ```

- file `test.lua`:

  ```lua
  -- Suppose we already installed casbin with "sudo luarocks install casbin"
  local casbin = require("casbin")
  local enforcer = casbin:new("model.conf", "policy.csv")
  
  -- Test
  print(enforcer:enforce("test", "/test/aB3", "GET"))
  ```

Result:

```lang-none
INFO    Model:
INFO    [e.e]:
INFO    {RM = {}, key = "e", policy = {}, policyMap = {}, priorityIndex = -1, tokens = {}, value = "some(where (p_eft == allow))"}
INFO    [p.p]:
INFO    {RM = {}, key = "p", policy = {}, policyMap = {}, priorityIndex = -1, tokens = {"p_sub", "p_obj", "p_act"}, value = "sub, obj, act"}
INFO    [m.m]:
INFO    {RM = {}, key = "m", policy = {}, policyMap = {}, priorityIndex = -1, tokens = {}, value = "regexMatch(r_sub, p_sub) && regexMatch(r_obj, p_obj) && regexMatch(r_act, p_act)"}
INFO    [r.r]:
INFO    {RM = {}, key = "r", policy = {}, policyMap = {}, priorityIndex = -1, tokens = {"r_sub", "r_obj", "r_act"}, value = "sub, obj, act"}
INFO    Policy:

INFO    p:   sub, obj, act:
INFO    {{"test", "^/test/[a-zA-Z0-9]{1", "3}$", "GET"}}
lua: /usr/local/share/lua/5.1/src/main/CoreEnforcer.lua:423: invalid policy size
stack traceback:
        [C]: in function 'error'
        /usr/local/share/lua/5.1/src/main/CoreEnforcer.lua:423: in function 'enforceEx'
        /usr/local/share/lua/5.1/src/main/CoreEnforcer.lua:534: in function 'enforce'
        test-regex.lua:6: in main chunk
        [C]: ? 
```

## What I Tried

I tried to **wrap the argument with double quotes**, as suggested on the documentation ([casbin.org/docs/policy-storage](https://casbin.org/docs/policy-storage/#loading-policy-from-a-csv-file)):
> NOTE
> If your file contains commas, you should **_wrap them in double quotes_**. For example:
> ```
> p, alice, "data1,data2", read    --correct
> p, alice, data1,data2, read        --incorrect (the whole phrase "data1,data2" should be wrapped in double quotes)
> ```

But with no success, since the double quotes are escaped:

```lang-none
INFO    {{"test", "\"^/test/[a-zA-Z0-9]{1", "3}$\"", "GET"}}
lua: /usr/local/share/lua/5.1/src/main/CoreEnforcer.lua:423: invalid policy size
```

## Online Editor

I also tried this with the online editor (based on [node-casbin v5.30.0](https://github.com/casbin/node-casbin/releases/tag/v5.30.0)), and appearently it works with double quotes: https://editor.casbin.org/#EB3SB7NZQ

- policy:
  
  ```csv
  p, test_no_quotes, ^/test/[a-zA-Z0-9]{1,3}$, GET
  p, test_quotes, "^/test/[a-zA-Z0-9]{1,3}$", GET
  ```

- request:
  
  ```lang-none
  test_no_quotes, /test/aB3, GET    # false
  test_quotes, /test/aB3, GET       # true Reason: ["test_quotes","^/test/[a-zA-Z0-9]{1,3}$","GET"]
  test_quotes, /test/aB3asd, GET    # false
  ```

---

Related issue: https://github.com/casbin/casbin/issues/886

### PR DESCRIPTION
**Pull request by mikyll**
Added Util.splitCSVLine() function that splits a CSV-formatted string into an array of tokens, accounting for quotes (single and double) and escape characters.

### CODE REVIEW

**Comment by hsluoyz** [comment]
@mikyll fix CI errors:

image

File: src/util/Util.lua [hunk 1]
```diff
@@ -252,4 +252,66 @@ function Util.printTable(t)
...
+    end
+  
+    -- Add the last field (since it won't have the delimiter after it)
+    if trimFields then
+        field = Util.trim(field)
+    end
+    table.insert(result, field)
+  
+    return result
+end
```
Comment (leeqvip) [inline src/util/Util.lua:315]: Please add some unit tests

**Review by leeqvip — CHANGES_REQUESTED** [review]

**Comment by mikyll** [comment]
@leeqvip I made some fixes to Util functions and added unit tests for a use case about #157 (`{N,M}` regex quantifier)

**Comment by mikyll** [comment]
# Summary

## Differences

### Old `Util.split(str, delimiter, x)` function

This function was used for simply splitting a CSV-formatted line, into an array of fields, separated by `delimiter` (up to x number of fields). Fields were also trimmed to remove leading/trailing whitespaces.

### New `Util.splitEnhanced(str, delimiter, trimFields)` function

This function does the same as `Util.split()`, but it adds the following features:
1. **quoted fields** - fields can be wrapped in double quotes (`"`) to escape comma characters that are part of the field (useful for example in regexMatch, to handle `{N,M}` pattern quantifiers);
2. **double quotes in quoted fields** - quoted fields can contain double quotes, which are escaped through another double quotes character (`""`).

    This complies with what is stated in the documentation ([Policy Storage | Casbin](https://casbin.org/docs/policy-storage)):
    
    <details>
    <summary>Show/Hide</summary>
    
    > **NOTE**
    > If your file contains commas, you should wrap them in double quotes. For example:
    > 
    > ```csv
    > p, alice, "data1,data2", read    --correct
    > p, alice, data1,data2, read        --incorrect (the whole phrase "data1,data2" should be wrapped in double quotes)
    > ```
    > 
    > If your file contains commas and double quotes, you should enclose the field in double quotes and double any embedded double quotes.
    > 
    > ```csv
    > p, alice, data, "r.act in (""get"", ""post"")"        --correct
    > p, alice, data, "r.act in ("get", "post")"            --incorrect (you should use "" to escape "")
    > ```
    > 
    > Related issue: [casbin#886](https://github.com/casbin/casbin/issues/886)

</details>

3. **policy syntax errors**:
    - if a quoted field presents extra trailing chracters, after the closing double quotes, the function throws the error `Quoted fields cannot have extra characters outside double quotes.`
    - if there are unmatched quotes, the function throws the error `Unmatched quotes.`

## New Example

- file `examples/basic_model_with_regex.conf` - implements a simple ACL model with regex for each field of the policy rules;
- file `examples/basic_policy_with_regex.csv` - shows a simple usage of this model with a rule that includes the pattern quantifier `{N,M}` (it includes a comma escaped through the **quoted field**, so it's a way to show the correct behaviour of the `Util.splitEnhanced()` function).

## Tests

### `util_spec`

- test **isOnlyWhitespaces** - verifies the correct behaviour of `Util.isOnlyWhitespaces()`;
- test **splitEnhanced** - verifies the correct behaviour of `Util.splitEnhanced()`. In particular, the usage of quotedFields, escaped double quotes and errors;

### `enforcer_spec`

- "regexMatch test" verifies the regexMatch behaviour of the new example, which exploits the new "quoted fields" feature.

**Review by leeqvip — APPROVED** [review]

**Pull request merged**